From 026479c00723fb86bf0a064e34011115dcf98318 Mon Sep 17 00:00:00 2001
From: gao.cui <gao.cui@windsolve.com>
Date: Sun, 13 Jul 2014 18:05:31 +0800
Subject: [PATCH 18/30] addition of SDHC driver for Curie
 SDHC2 is used for the Wi-Fi module
 SDHC3 is used for the SD Card slot
 SDHC4 is used for the on-board eMMC

Wi-Fi module works at 4-bit mode
SD Card requires another 2 GPIOs for card-detection and write-protection which are GPIO(6,8) and GPIO(6,15)
---
 arch/arm/mach-mx6/board-mx6q_curie.c |   38 ++++++++++++++++++++++++++++++++++++
 arch/arm/mach-mx6/board-mx6q_curie.h |   35 +++++++++++++++++++++++++++++++++
 2 files changed, 73 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-mx6/board-mx6q_curie.c b/arch/arm/mach-mx6/board-mx6q_curie.c
index c5bc4c2..59e77c3 100644
--- a/arch/arm/mach-mx6/board-mx6q_curie.c
+++ b/arch/arm/mach-mx6/board-mx6q_curie.c
@@ -232,6 +232,37 @@ static struct platform_device mx6q_curie_vmmc_reg_devices = {
 	},
 };
 
+/* SDHC2 for Wi-Fi Module */
+static const struct esdhc_platform_data mx6q_curie_sd2_data __initconst = {
+	.always_present = 1,
+	.keep_power_at_suspend = 1,
+	.support_8bit = 1,
+	.delay_line = 0,
+	.cd_type = ESDHC_CD_PERMANENT,
+};
+
+/* SDHC3 for SD Card Slot */
+#define CURIE_SD3_CD        IMX_GPIO_NR(6, 8)
+#define CURIE_SD3_WP        IMX_GPIO_NR(6, 15)
+static const struct esdhc_platform_data mx6q_curie_sd3_data __initconst = {
+	.cd_gpio = CURIE_SD3_CD,
+	.wp_gpio = CURIE_SD3_WP,
+	.keep_power_at_suspend = 1,
+	.support_8bit = 1,
+	.delay_line = 0,
+	.cd_type = ESDHC_CD_CONTROLLER,
+};
+
+/* SDHC4 for eMMC */
+static const struct esdhc_platform_data mx6q_curie_sd4_data __initconst = {
+	.always_present = 1,
+	.keep_power_at_suspend = 1,
+	.support_8bit = 1,
+	.delay_line = 0,
+	.cd_type = ESDHC_CD_PERMANENT,
+};
+
+
 /* Board Functions */
 static void __init fixup_mxc_board(struct machine_desc *desc, struct tag *tags,
 				   char **cmdline, struct meminfo *mi)
@@ -282,6 +313,13 @@ static void __init mx6_curie_board_init(void)
 	imx6q_add_anatop_thermal_imx(1, &mx6q_curie_anatop_thermal_data);
 	/* VMMC Regulator */
 	platform_device_register(&mx6q_curie_vmmc_reg_devices);
+	/* SDHC
+	   Move sd4 to first because sd4 connect to emmc.
+	   Mfgtools want emmc is mmcblk0 and other sd card is mmcblk1.
+	 */
+	imx6q_add_sdhci_usdhc_imx(3, &mx6q_curie_sd4_data);
+	imx6q_add_sdhci_usdhc_imx(1, &mx6q_curie_sd2_data);
+	imx6q_add_sdhci_usdhc_imx(2, &mx6q_curie_sd3_data);
 }
 
 extern void __iomem *twd_base;
diff --git a/arch/arm/mach-mx6/board-mx6q_curie.h b/arch/arm/mach-mx6/board-mx6q_curie.h
index 6b9d053..e539137 100644
--- a/arch/arm/mach-mx6/board-mx6q_curie.h
+++ b/arch/arm/mach-mx6/board-mx6q_curie.h
@@ -60,6 +60,41 @@ static iomux_v3_cfg_t mx6q_curie_pads[] = {
 
 	/* PMIC_INT_B */
 	MX6Q_PAD_GPIO_18__GPIO_7_13,
+
+	/* USDHC2 */
+	MX6Q_PAD_SD2_CLK__USDHC2_CLK,
+	MX6Q_PAD_SD2_CMD__USDHC2_CMD,
+	MX6Q_PAD_SD2_DAT0__USDHC2_DAT0,
+	MX6Q_PAD_SD2_DAT1__USDHC2_DAT1,
+	MX6Q_PAD_SD2_DAT2__USDHC2_DAT2,
+	MX6Q_PAD_SD2_DAT3__USDHC2_DAT3,
+
+	/* USDHC3 */
+	MX6Q_PAD_SD3_CLK__USDHC3_CLK_50MHZ,
+	MX6Q_PAD_SD3_CMD__USDHC3_CMD_50MHZ,
+	MX6Q_PAD_SD3_DAT0__USDHC3_DAT0_50MHZ,
+	MX6Q_PAD_SD3_DAT1__USDHC3_DAT1_50MHZ,
+	MX6Q_PAD_SD3_DAT2__USDHC3_DAT2_50MHZ,
+	MX6Q_PAD_SD3_DAT3__USDHC3_DAT3_50MHZ,
+	MX6Q_PAD_SD3_DAT4__USDHC3_DAT4_50MHZ,
+	MX6Q_PAD_SD3_DAT5__USDHC3_DAT5_50MHZ,
+	MX6Q_PAD_SD3_DAT6__USDHC3_DAT6_50MHZ,
+	MX6Q_PAD_SD3_DAT7__USDHC3_DAT7_50MHZ,
+	MX6Q_PAD_NANDF_ALE__GPIO_6_8,		/* SD3_CD */
+	MX6Q_PAD_NANDF_CS2__GPIO_6_15,		/* SD3_WP */
+
+	/* USDHC4 */
+	MX6Q_PAD_SD4_CLK__USDHC4_CLK_50MHZ,
+	MX6Q_PAD_SD4_CMD__USDHC4_CMD_50MHZ,
+	MX6Q_PAD_SD4_DAT0__USDHC4_DAT0_50MHZ,
+	MX6Q_PAD_SD4_DAT1__USDHC4_DAT1_50MHZ,
+	MX6Q_PAD_SD4_DAT2__USDHC4_DAT2_50MHZ,
+	MX6Q_PAD_SD4_DAT3__USDHC4_DAT3_50MHZ,
+	MX6Q_PAD_SD4_DAT4__USDHC4_DAT4_50MHZ,
+	MX6Q_PAD_SD4_DAT5__USDHC4_DAT5_50MHZ,
+	MX6Q_PAD_SD4_DAT6__USDHC4_DAT6_50MHZ,
+	MX6Q_PAD_SD4_DAT7__USDHC4_DAT7_50MHZ,
+
 };
 
 #endif
-- 
1.7.0.4

